watchdog/index.js

```diff
@@ -45,24 +45,25 @@ class Watchdog {
     }
     privateKey = Buffer.from(privateKey, 'hex')
     while (true) {
-      const lastClientBlock = await this.clientContract.methods.head().call()
-      const latestBlock = await this.robustWeb3.getBlock('latest')
-      console.log(
-        `Examining block ${lastClientBlock.hash} height: ${lastClientBlock.height}`
-      )
-      if (latestBlock.timestamp >= lastClientBlock.valid) {
-        const timeDelta = 10
-        console.log(`Block is valid. Sleeping for ${timeDelta} seconds.`)
-        await sleep(timeDelta * 1000)
+      const bridgeState = await this.clientContract.methods.bridgeState().call()
+      if (Number(bridgeState.nextValidAt) == 0) {
+        const delay = 10
+        console.log(`No block to check. Sleeping for ${delay} seconds.`)
+        await sleep(delay * 1000)
         continue
       }
 
+      console.log('Checking block.')
       // We cannot memorize processed blocks because they might have been re-submitted with different data.
-      for (let i = 0; i < lastClientBlock.approvals_after_next_length; i++) {
+      for (let i = 0; i < Number(bridgeState.numBlockProducers); i++) {
         console.log(`Checking signature ${i}.`)
-        const result = await this.clientContract.methods
-          .checkBlockProducerSignatureInHead(i)
-          .call()
+        let result
+        try {
+          result = await this.clientContract.methods.checkBlockProducerSignatureInHead(i).call()
+        } catch (e) {
+          console.log(`Error, continuing.`)
```

Comment:
I suggest to not attempt to call `checkBlockProducerSignatureInHead` when signature is missing, because currently we cannot differentiate between transient errors, e.g. RPC-related stuff and error that occurs when we try to call `checkBlockProducerSignatureInHead` on missing signature.

Reply:
The current logic should work fine in both cases. Suppose that it will be possible to differentiate between those cases, will watchdog's behavior be different?

Reply:
Right now sleeps `timeDelta` before revalidating the block. Ideally this should be parameter (currently when I run on gcloud I patch this code for `timeDelta` to be `300` because otherwise we would exceed infura quota). If user launches watchdog with `timeDelta` that is comparable to challenge period then watchdog can accidentally skip invalid signature because of some RPC error and wait to retry for `timeDelta`, which might be sufficient for bad header to be accepted. If we can differentiate between these two errors then we can retry signature check until RPC succeeds and only then sleep for `timeDelta`.

Also, currently this will print lots of spam logs, because testnet headers have routinely skipped signatures. So right now `watchdog` will be printing several `Error, continuing` every `timeDelta` seconds, which might be misinterpreted by the users.

Reply:
I made the delay configurable, so this shouldn't be an issue anymore.